core/serialize/serialize.cc

```diff
@@ -973,24 +973,24 @@ Pickler SerializerImpl::pickleNameTable(const GlobalState &gs) {
 }
 
 void SerializerImpl::pickleNameTable(Pickler &p, const GlobalState &gs) {
+    // Write the size of the hash table out first, so that we can pre-allocate it when deserializing.
+    p.putU4(gs.namesByHash.size());
```

Comment:
It might be better to write the sizes of the three tables first, and then derive the size of the hash table from that.